File: php/limport.php
#!/usr/bin/env php
<?php

require_once __DIR__ . '/ledger.php';

use LedgerPHP\Ledger;
use LedgerPHP\Parser;
use LedgerPHP\SimpleRational;

define('TRANSACTION_DATE_FORMAT', 'Y/m/d');
define('DISPLAY_PRECISION', 2);

class LimportCLI
{
    private $options = [
        'f' => '',
        'neg' => false,
        'allow-matching' => false,
        'scale' => 1.0,
        'set-search' => 'Expenses',
        'date-format' => 'm/d/Y',
        'delimiter' => ',',
        'columns' => 79,
        'wide' => false,
    ];
    
    private $account;
    private $csvFileName;

    public function run(array $argv): void
    {
        $this->parseArguments($argv);
        
        if (empty($this->options['f']) || empty($this->account) || empty($this->csvFileName)) {
            $this->showUsage();
            exit(1);
        }
        
        if ($this->options['wide']) {
            $this->options['columns'] = 132;
        }
        
        try {
            // Load ledger
            $ledgerContent = file_get_contents($this->options['f']);
            if ($ledgerContent === false) {
                throw new \RuntimeException("Could not read ledger file");
            }
            
            $generalLedger = Parser::parseLedger($ledgerContent);
            
            // FIND DESTINATION ACCOUNT
            $this->account = $this->findDestinationAccount($generalLedger, $this->account);
            if ($this->account === null) {
                throw new \RuntimeException("Unable to find matching account.");
            }
            
            // BUILD BAYESIAN CLASSIFIER (same as Go)
            $classifier = $this->buildBayesianClassifier($generalLedger);
            
            // Process CSV
            $this->processCSV($generalLedger, $classifier);
            
        } catch (\Exception $e) {
            echo "Error: " . $e->getMessage() . "\n";
            exit(1);
        }
    }
    
    private function findDestinationAccount(array $transactions, string $accountSubstring): ?string
    {
        $allAccounts = [];
        foreach ($transactions as $transaction) {
            foreach ($transaction->accountChanges as $accountChange) {
                $allAccounts[$accountChange->name] = true;
            }
        }
        
        $matchingAccounts = [];
        foreach (array_keys($allAccounts) as $accountName) {
            if (stripos($accountName, $accountSubstring) !== false) {
                $matchingAccounts[] = $accountName;
            }
        }
        
        if (empty($matchingAccounts)) {
            return null;
        }
        
        return $matchingAccounts[count($matchingAccounts) - 1];
    }
    
    private function buildBayesianClassifier(array $transactions): array
    {
        // First, collect all accounts that contain the search string
        $classes = [];
        
        foreach ($transactions as $transaction) {
            foreach ($transaction->accountChanges as $accountChange) {
                if (stripos($accountChange->name, $this->options['set-search']) !== false) {
                    $classes[$accountChange->name] = true;
                }
            }
        }
        
        $classes = array_keys($classes);
        
        // Initialize classifier
        $classifier = [
            'classes' => $classes,
            'datas' => [],
            'learned' => 0
        ];
        
        foreach ($classes as $class) {
            $classifier['datas'][$class] = [
                'freqs' => [],
                'total' => 0
            ];
        }
        
        // Train the classifier - same as Go
        foreach ($transactions as $transaction) {
            $payeeWords = $this->extractWords($transaction->payee);
            
            foreach ($transaction->accountChanges as $accountChange) {
                $accountName = $accountChange->name;
                
                // Only consider accounts that are in the classes
                if (in_array($accountName, $classes)) {
                    foreach ($payeeWords as $word) {
                        if (!isset($classifier['datas'][$accountName]['freqs'][$word])) {
                            $classifier['datas'][$accountName]['freqs'][$word] = 0;
                        }
                        $classifier['datas'][$accountName]['freqs'][$word]++;
                        $classifier['datas'][$accountName]['total']++;
                    }
                    $classifier['learned']++;
                }
            }
        }
        
        return $classifier;
    }
    
    private function extractWords(string $text): array
    {
        // SAME AS GO: simply splits by spaces, keeps all words
        $text = strtolower(trim($text));
        $words = preg_split('/\s+/', $text, -1, PREG_SPLIT_NO_EMPTY);
        return $words ?: [];
    }
    
    private function getWordProb(array $classData, string $word): float
    {
        // SAME AS GO: P(W|C) = (count(W,C) + 1) / (total_words_in_C + vocabulary_size)
        $vocabSize = count($classData['freqs']);
        if ($classData['total'] == 0 || $vocabSize == 0) {
            return 1e-11; // defaultProb from Go
        }
        
        $value = $classData['freqs'][$word] ?? 0;
        return ($value + 1) / ($classData['total'] + $vocabSize);
    }
    
    private function getPriors(array $classifier): array
    {
        // SAME AS GO: P(C_j) = (count_j + 1) / (total + num_classes)
        $n = count($classifier['classes']);
        $priors = [];
        $sum = 0;
        
        foreach ($classifier['classes'] as $index => $class) {
            $total = $classifier['datas'][$class]['total'];
            $priors[$index] = $total;
            $sum += $total;
        }
        
        // Apply Laplace smoothing
        $floatN = (float)$n;
        $floatSum = (float)$sum;
        foreach ($priors as $index => $prior) {
            $priors[$index] = ($prior + 1) / ($floatSum + $floatN);
        }
        
        return $priors;
    }
    
    private function logScores(array $classifier, array $document): array
    {
        // SAME AS GO: classifier.LogScores()
        $n = count($classifier['classes']);
        $scores = array_fill(0, $n, 0.0);
        $priors = $this->getPriors($classifier);
        
        foreach ($classifier['classes'] as $index => $class) {
            $data = $classifier['datas'][$class];
            $score = log($priors[$index]);
            
            foreach ($document as $word) {
                $wordProb = $this->getWordProb($data, $word);
                $score += log($wordProb);
            }
            
            $scores[$index] = $score;
        }
        
        return $scores;
    }
    
    private function findMax(array $scores): array
    {
        // SAME AS GO: findMax function
        $inx = 0;
        $strict = true;
        
        for ($i = 1; $i < count($scores); $i++) {
            if ($scores[$inx] < $scores[$i]) {
                $inx = $i;
                $strict = true;
            } elseif ($scores[$inx] == $scores[$i]) {
                $strict = false;
            }
        }
        
        return ['inx' => $inx, 'strict' => $strict];
    }
    
    private function classifyPayee(array $classifier, string $payee): string
    {
        $words = $this->extractWords($payee);
        
        if (empty($words) || empty($classifier['classes'])) {
            return 'unknown:unknown';
        }
        
        $scores = $this->logScores($classifier, $words);
        $result = $this->findMax($scores);
        
        return $classifier['classes'][$result['inx']];
    }
    
    private function parseArguments(array $argv): void
    {
        $options = [];
        $positionalArgs = [];
        
        $i = 1;
        while ($i < count($argv)) {
            $arg = $argv[$i];
            
            if ($arg === '--help') {
                $this->showUsage();
                exit(0);
            }
            
            if (strpos($arg, '--') === 0) {
                if (strpos($arg, '=') !== false) {
                    $parts = explode('=', substr($arg, 2), 2);
                    $key = $parts[0];
                    $value = $parts[1];
                    $options[$key] = $value;
                } else {
                    $key = substr($arg, 2);
                    if ($i + 1 < count($argv) && strpos($argv[$i + 1], '-') !== 0) {
                        $options[$key] = $argv[$i + 1];
                        $i++;
                    } else {
                        $options[$key] = true;
                    }
                }
            } elseif (strpos($arg, '-') === 0 && strlen($arg) == 2) {
                $key = $arg[1];
                if ($i + 1 < count($argv) && strpos($argv[$i + 1], '-') !== 0) {
                    $options[$key] = $argv[$i + 1];
                    $i++;
                } else {
                    $options[$key] = true;
                }
            } else {
                $positionalArgs[] = $arg;
            }
            
            $i++;
        }
        
        $this->options = array_merge($this->options, $options);
        
        if (count($positionalArgs) >= 2) {
            $this->account = $positionalArgs[0];
            $this->csvFileName = $positionalArgs[1];
        } else {
            echo "Error: Insufficient arguments.\n";
            $this->showUsage();
            exit(1);
        }
    }
    
    private function processCSV(array $generalLedger, array $classifier): void
    {
        if (!file_exists($this->csvFileName)) {
            throw new \RuntimeException("CSV file not found: " . $this->csvFileName);
        }
        
        $csvContent = file_get_contents($this->csvFileName);
        if ($csvContent === false) {
            throw new \RuntimeException("Could not read CSV file");
        }
        
        $lines = explode("\n", trim($csvContent));
        $header = str_getcsv(array_shift($lines), $this->options['delimiter'], '"', '\\');
        
        $dateColumn = $payeeColumn = $amountColumn = $commentColumn = $uuidColumn = $buyerColumn = -1;
        
        foreach ($header as $index => $fieldName) {
            $fieldName = strtolower(trim($fieldName));
            if (strpos($fieldName, 'date') !== false) {
                $dateColumn = $index;
            } elseif (strpos($fieldName, 'description') !== false || strpos($fieldName, 'payee') !== false) {
                $payeeColumn = $index;
            } elseif (strpos($fieldName, 'amount') !== false || strpos($fieldName, 'expense') !== false) {
                $amountColumn = $index;
            } elseif (strpos($fieldName, 'note') !== false) {
                $commentColumn = $index;
            } elseif (strpos($fieldName, 'uuid') !== false) {
                $uuidColumn = $index;
            } elseif (strpos($fieldName, 'buyer') !== false) {
                $buyerColumn = $index;
            }
        }
        
        if ($dateColumn < 0 || $payeeColumn < 0 || $amountColumn < 0) {
            throw new \RuntimeException("Unable to find columns required from header field names.");
        }
        
        foreach ($lines as $line) {
            if (trim($line) === '') continue;
            
            $record = str_getcsv($line, $this->options['delimiter'], '"', '\\');
            $record = array_pad($record, max($dateColumn, $payeeColumn, $amountColumn, $commentColumn, $uuidColumn, $buyerColumn) + 1, '');
            
            $dateStr = trim($record[$dateColumn]);
            $date = $this->parseDate($dateStr);
            
            if ($date === false) {
                continue;
            }
            
            $payee = trim($record[$payeeColumn]);
            $payeeWords = explode(' ', $payee);
            
            if (!$this->options['allow-matching'] && $this->existingTransaction($generalLedger, $date, $payeeWords[0] ?? '')) {
                continue;
            }
            
            // Classify using EXACTLY the same algorithm as Go
            $classifiedAccount = $this->classifyPayee($classifier, $payee);
            
            // Parse value
            $amountStr = trim($record[$amountColumn]);
            $amount = $this->parseAmount($amountStr);
            
            if ($amount === null) {
                continue;
            }
            
            $amount *= (float)$this->options['scale'];
            
            // CORREÇÃO: Se --neg estiver ativo, inverte o sinal do amount original
            if ($this->options['neg']) {
                $amount = -$amount;
            }
            
            // Logic from Go
            $csvAmount = -$amount;
            $expenseAmount = -$csvAmount;
            
            // Comments - REMOVE extra space after semicolon
            $comments = [];
            if ($commentColumn >= 0 && trim($record[$commentColumn]) !== '') {
                $comments[] = ';' . trim($record[$commentColumn]); // NO space
            }
            if ($uuidColumn >= 0 && trim($record[$uuidColumn]) !== '') {
                $comments[] = '; UUID: ' . trim($record[$uuidColumn]);
            }
            if ($buyerColumn >= 0 && trim($record[$buyerColumn]) !== '') {
                $comments[] = '; Buyer: ' . trim($record[$buyerColumn]);
            }
            
            // Print transaction
            $this->printTransaction($date, $payee, $this->account, $csvAmount, $classifiedAccount, $expenseAmount, $comments);
        }
    }
    
    private function parseDate(string $dateStr)
    {
        // Remove espaços extras
        $dateStr = trim($dateStr);
        
        // Se temos um formato específico configurado
        if (!empty($this->options['date-format'])) {
            $format = $this->options['date-format'];
            
            // Tentar o formato configurado
            $date = \DateTime::createFromFormat($format, $dateStr);
            if ($date !== false) {
                return $date;
            }
            
            // Tentar variantes com diferentes separadores
            $variants = [
                str_replace('/', '-', $format),
                str_replace('-', '/', $format),
                str_replace('/', '.', $format),
                str_replace('.', '/', $format),
            ];
            
            foreach ($variants as $variant) {
                if ($variant !== $format) {
                    $date = \DateTime::createFromFormat($variant, $dateStr);
                    if ($date !== false) {
                        return $date;
                    }
                }
            }
        }
        
        // Lista mais abrangente de formatos (em ordem de prioridade para Brasil)
        $formats = [
            // Formatos brasileiros primeiro
            'd/m/Y',      // 31/12/2023
            'd-m-Y',      // 31-12-2023  
            'd.m.Y',      // 31.12.2023
            
            // Formatos ISO/internacionais
            'Y-m-d',      // 2023-12-31
            'Y/m/d',      // 2023/12/31
            'Y.m.d',      // 2023.12.31
            
            // Formatos americanos
            'm/d/Y',      // 12/31/2023
            'm-d-Y',      // 12-31-2023
            'm.d.Y',      // 12.31.2023
            
            // Com barras invertidas
            'd\\m\\Y',    // 31\12\2023
            'Y\\m\\d',    // 2023\12\31
            
            // Com hora (ignora a hora)
            'Y-m-d H:i:s',    // 2023-12-31 14:30:00
            'd/m/Y H:i:s',    // 31/12/2023 14:30:00
            'm/d/Y H:i:s',    // 12/31/2023 14:30:00
            'Y-m-d\TH:i:s',   // 2023-12-31T14:30:00 (ISO com T)
            
            // Formatos com dia/mês com 1 dígito
            'j/n/Y',      // 31/12/2023 (sem zero à esquerda)
            'j-n-Y',      // 31-12-2023
            'n/j/Y',      // 12/31/2023 (EUA sem zero)
            
            // Datas apenas numéricas (auto-detectar)
            'Ymd',        // 20231231
            'dmY',        // 31122023
            'mdy',        // 12312023
        ];
        
        foreach ($formats as $format) {
            $date = \DateTime::createFromFormat($format, $dateStr);
            if ($date !== false) {
                // Verificar se a data é válida (não é falsa como 31/02/2023)
                $errors = \DateTime::getLastErrors();
                // getLastErrors() pode retornar false se não houver erros
                if ($errors === false || ($errors['warning_count'] === 0 && $errors['error_count'] === 0)) {
                    return $date;
                }
            }
        }
        
        // Se não funcionou com os formatos acima, tentar detecção inteligente
        return $this->parseDateIntelligent($dateStr);
    }
    
    private function parseAmount(string $amountStr): ?float
    {
        $amountStr = trim($amountStr);
        
        if (preg_match('/^\((.+)\)$/', $amountStr, $matches)) {
            $amountStr = '-' . $matches[1];
        }
        
        $amountStr = preg_replace('/[^\d\.,\-]/', '', $amountStr);
        
        if ($amountStr === '') {
            return 0.0;
        }
        
        $negative = (substr_count($amountStr, '-') % 2) == 1;
        $amountStr = str_replace('-', '', $amountStr);
        
        $amountStr = str_replace(',', '.', $amountStr);
        
        if (!is_numeric($amountStr)) {
            return null;
        }
        
        $amount = (float)$amountStr;
        if ($negative) {
            $amount = -$amount;
        }
        
        return $amount;
    }
    
    private function existingTransaction(array $generalLedger, \DateTime $transDate, string $payee): bool
    {
        foreach ($generalLedger as $trans) {
            if ($trans->date->format('Y-m-d') === $transDate->format('Y-m-d')) {
                if (strpos($trans->payee, $payee) === 0) {
                    return true;
                }
            }
        }
        return false;
    }
    
    private function printTransaction(\DateTime $date, string $payee, string $csvAccount, 
                                     float $csvAmount, string $expenseAccount, 
                                     float $expenseAmount, array $comments): void
    {
        foreach ($comments as $comment) {
            echo $comment . "\n";
        }
        
        echo $date->format(TRANSACTION_DATE_FORMAT) . " " . $payee . "\n";
        
        $this->printAccountLine($csvAccount, $csvAmount);
        $this->printAccountLine($expenseAccount, $expenseAmount);
        
        echo "\n";
    }
    
    private function printAccountLine(string $accountName, float $amount): void
    {
        $formattedValue = number_format($amount, DISPLAY_PRECISION, '.', '');
        
        $indent = str_repeat(' ', 4); // Always 4 spaces indent
        $nameLength = strlen($accountName);
        
        // Calculate available width for name and value
        $availableWidth = $this->options['columns'] - 4; // 4 spaces indent
        $valueWidth = 12; // Width for values (10 digits + 2 spaces)
        
        // If not enough space, truncate account name
        if ($nameLength > $availableWidth - $valueWidth - 2) {
            $maxDisplayLength = $availableWidth - $valueWidth - 5; // Leave space for "..."
            if ($maxDisplayLength > 10) {
                $accountName = substr($accountName, 0, $maxDisplayLength) . '...';
                $nameLength = strlen($accountName);
            }
        }
        
        // Calculate spaces to right align value
        $totalSpaces = $availableWidth - $nameLength - strlen($formattedValue);
        if ($totalSpaces < 2) {
            $totalSpaces = 2;
        }
        
        echo $indent . $accountName . str_repeat(' ', $totalSpaces) . $formattedValue . "\n";
    }
    
    private function showUsage(): void
    {
        echo "Limport - CSV Importer for Ledger (results only)\n";
        echo "=================================================\n\n";
        echo "Usage:\n";
        echo "  php limport.php -f <ledger-file> [options] <account> <csv-file>\n\n";
        echo "Options:\n";
        echo "  -f FILE          Ledger file (*required)\n";
        echo "  --neg            Negate amount column value\n";
        echo "  --allow-matching Include imported transactions that match existing ledger transactions\n";
        echo "  --scale=FACTOR   Scaling factor to multiply each imported value by (default: 1.0)\n";
        echo "  --set-search=STR Search string used to find account set for classification\n";
        echo "  --date-format=STR Date format (default: m/d/Y)\n";
        echo "  --delimiter=STR  Field delimiter (default: ,)\n";
        echo "  --columns=N      Column width (default: 79)\n";
        echo "  --wide           Wide mode (132 columns)\n";
        echo "  --help           Show this help\n\n";
        echo "Examples:\n";
        echo "  php limport.php -f Journal.txt Paypal paypal.csv\n";
        echo "  php limport.php -f Journal.txt --set-search Expenses Paypal paypal.csv\n";
        echo "  php limport.php -f Journal.txt --set-search Utilities Paypal paypal.csv\n";
        echo "  php limport.php -f Journal.txt --columns=132 Paypal paypal.csv\n";
        echo "  php limport.php -f Journal.txt --wide Paypal paypal.csv\n\n";
        echo "This script prints only imported transactions, without debug information.\n";
    }
}

if (PHP_SAPI === 'cli') {
    $cli = new LimportCLI();
    $cli->run($argv);
}

File: python/limport.py
#!/usr/bin/env python3

import csv
import re
import sys
import math
from datetime import datetime
from typing import List, Dict, Optional, Any, Tuple

TRANSACTION_DATE_FORMAT = "%Y/%m/%d"
DISPLAY_PRECISION = 2


class SimpleRational:
    """Simple rational number class for accounting calculations"""

    def __init__(self, value: float):
        self.value = value

    @classmethod
    def zero(cls):
        return cls(0.0)

    def plus(self, other):
        return SimpleRational(self.value + other.value)

    def is_zero(self):
        return abs(self.value) < 0.000001

    def to_f(self, precision=2):
        return f"{self.value:.{precision}f}"


class AccountChange:
    def __init__(self, name: str, balance: SimpleRational = None):
        self.name = name
        self.balance = balance


class Transaction:
    def __init__(self, payee: str, date: datetime):
        self.payee = payee
        self.date = date
        self.accountChanges = []
        self.comments = []


class Parser:
    @classmethod
    def parse_ledger(cls, content: str) -> List[Transaction]:
        transactions = []
        lines = content.split('\n')
        current = None
        comments = []

        for line in lines:
            line = line.rstrip()
            if not line.strip():
                if current:
                    current.comments = comments
                    transactions.append(current)
                    current = None
                    comments = []
                continue

            if line.strip().startswith(';'):
                comments.append(line)
                continue

            # Match date at beginning: YYYY/MM/DD
            match = re.match(r'^(\d{4}/\d{2}/\d{2})\s+(.+)$', line)
            if match:
                if current:
                    current.comments = comments
                    transactions.append(current)
                    comments = []
                date_str = match.group(1)
                payee = match.group(2)
                try:
                    date = datetime.strptime(date_str, "%Y/%m/%d")
                    current = Transaction(payee, date)
                except:
                    continue
            elif current and (line.startswith('    ') or line.startswith('\t')):
                line = line.strip()
                match2 = re.match(r'^(.+?)\s{2,}(.+)$', line)
                if match2:
                    name = match2.group(1).strip()
                    value_str = match2.group(2).strip().replace(',', '.')
                    try:
                        value = float(value_str)
                        current.accountChanges.append(AccountChange(name, SimpleRational(value)))
                    except:
                        pass
                else:
                    current.accountChanges.append(AccountChange(line))

        if current:
            current.comments = comments
            transactions.append(current)

        transactions.sort(key=lambda t: t.date)
        return transactions


def extract_words(text: str) -> List[str]:
    """Extract words from text, same as Go"""
    return text.lower().strip().split()


def get_word_prob(freqs: Dict[str, int], total: int, word: str, vocab_size: int) -> float:
    """Calculate P(W|C) = (count(W,C) + 1) / (total_words_in_C + vocabulary_size)"""
    if total == 0 or vocab_size == 0:
        return 1e-11
    value = freqs.get(word, 0)
    return (value + 1) / (total + vocab_size)


def get_priors(classes: List[str], datas: Dict[str, Dict[str, Any]]) -> List[float]:
    """Calculate priors P(C_j) = (count_j + 1) / (total + num_classes)"""
    n = len(classes)
    priors = []
    total_sum = 0
    for klass in classes:
        total = datas[klass]['total']
        priors.append(float(total))
        total_sum += total

    float_n = float(n)
    float_sum = float(total_sum)
    return [(p + 1) / (float_sum + float_n) for p in priors]


def log_scores(classes: List[str], datas: Dict[str, Dict[str, Any]], document: List[str]) -> List[float]:
    """Calculate log scores for each class"""
    n = len(classes)
    scores = [0.0] * n
    priors = get_priors(classes, datas)

    for idx, klass in enumerate(classes):
        data = datas[klass]
        freqs = data['freqs']
        total = data['total']
        vocab_size = len(freqs)

        if priors[idx] > 0:
            score = math.log(priors[idx])
        else:
            score = -float('inf')

        for word in document:
            word_prob = get_word_prob(freqs, total, word, vocab_size)
            score += math.log(word_prob)
        scores[idx] = score

    return scores


def find_max(scores: List[float]) -> Tuple[int, bool]:
    """Find index of maximum score - same as PHP/Go"""
    inx = 0
    strict = True
    
    for i in range(1, len(scores)):
        if scores[inx] < scores[i]:
            inx = i
            strict = True
        elif scores[inx] == scores[i]:
            strict = False
    
    return inx, strict


def classify_payee(payee: str, classes: List[str], datas: Dict[str, Dict[str, Any]]) -> str:
    """Classify payee using Bayesian classifier"""
    words = extract_words(payee)
    if not words or not classes:
        return "unknown:unknown"

    scores = log_scores(classes, datas, words)
    inx, _ = find_max(scores)
    return classes[inx]


def print_account_line(name: str, amount: float, columns: int) -> str:
    """Print account line with right-aligned amount at 79 columns (PHP style)"""
    amount_str = f"{amount:.{DISPLAY_PRECISION}f}"
    # Calculate spaces needed to right-align at column 79
    # PHP uses: spaces = columns - 4 - name_length - amount_length
    spaces = columns - 4 - len(name) - len(amount_str)
    if spaces < 2:
        spaces = 2
    return f"    {name}{' ' * spaces}{amount_str}"


def main():
    import argparse
    import math

    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-f', dest='file', default='')
    parser.add_argument('--set-search', dest='set_search', default='Expenses')
    parser.add_argument('--neg', dest='neg', action='store_true', default=False)
    parser.add_argument('--allow-matching', dest='allow_matching', action='store_true', default=False)
    parser.add_argument('--scale', dest='scale', type=float, default=1.0)
    parser.add_argument('--date-format', dest='date_format', default='m/d/Y')
    parser.add_argument('--delimiter', dest='delimiter', default=',')
    parser.add_argument('--columns', dest='columns', type=int, default=79)
    parser.add_argument('--wide', dest='wide', action='store_true', default=False)
    parser.add_argument('--help', dest='help', action='store_true', default=False)
    parser.add_argument('account', nargs='?', default='')
    parser.add_argument('csv_file', nargs='?', default='')

    args = parser.parse_args()

    if args.help:
        print("Limport - CSV Importer for Ledger")
        print("=================================")
        print()
        print("Usage: python limport.py -f <ledger-file> [options] <account> <csv-file>")
        print("Options:")
        print("  --set-search=STR   Search string for classification (default: Expenses)")
        print("  --neg              Negate amount column value")
        print("  --allow-matching   Include transactions that match existing ones")
        print("  --scale=FACTOR     Scaling factor (default: 1.0)")
        print("  --date-format=STR  Date format (default: m/d/Y)")
        print("  --delimiter=STR    Field delimiter (default: ,)")
        print("  --columns=N        Column width (default: 79)")
        print("  --wide             Wide mode (132 columns)")
        return

    if not args.file:
        print("Error: Ledger file required (-f)")
        return

    if not args.account or not args.csv_file:
        print("Error: Account and CSV file required")
        print("Usage: python limport.py -f <ledger-file> <account> <csv-file>")
        return

    if args.wide:
        args.columns = 132

    try:
        # Read ledger
        with open(args.file, 'r', encoding='utf-8') as f:
            content = f.read()
        transactions = Parser.parse_ledger(content)

        # Find destination account
        all_accounts = set()
        for t in transactions:
            for ac in t.accountChanges:
                all_accounts.add(ac.name)

        matching = [name for name in all_accounts if args.account.lower() in name.lower()]
        if not matching:
            print(f"Error: Account '{args.account}' not found")
            return
        dest_account = matching[-1]

        # Build Bayesian classifier - preserve order as they appear in ledger
        classes = []
        for t in transactions:
            for ac in t.accountChanges:
                if args.set_search.lower() in ac.name.lower() and ac.name not in classes:
                    classes.append(ac.name)

        if not classes:
            print(f"Warning: No classes found containing '{args.set_search}'")
            classes = [args.set_search]

        # Initialize classifier data
        datas = {}
        for klass in classes:
            datas[klass] = {'freqs': {}, 'total': 0}

        # Train the classifier - same as Go
        learned = 0
        for t in transactions:
            payee_words = extract_words(t.payee)
            for ac in t.accountChanges:
                if ac.name in classes:
                    data = datas[ac.name]
                    freqs = data['freqs']
                    for word in payee_words:
                        freqs[word] = freqs.get(word, 0) + 1
                    datas[ac.name] = {'freqs': freqs, 'total': data['total'] + 1}
                    learned += 1

        # Read CSV
        with open(args.csv_file, 'r', encoding='utf-8') as f:
            csv_content = f.read()

        lines = [l.strip() for l in csv_content.split('\n') if l.strip()]
        header = lines[0].split(args.delimiter)

        # Find column indices
        date_col = -1
        payee_col = -1
        amount_col = -1
        note_col = -1

        for idx, h in enumerate(header):
            h_lower = h.lower()
            if 'date' in h_lower:
                date_col = idx
            elif 'description' in h_lower or 'payee' in h_lower:
                payee_col = idx
            elif 'amount' in h_lower or 'expense' in h_lower:
                amount_col = idx
            elif 'note' in h_lower or 'comment' in h_lower:
                note_col = idx

        if date_col < 0 or payee_col < 0 or amount_col < 0:
            print("Error: Required columns not found")
            return

        # Process CSV lines
        for line in lines[1:]:
            record = line.split(args.delimiter)
            if len(record) <= max(date_col, payee_col, amount_col):
                continue

            date_str = record[date_col].strip()
            payee = record[payee_col].strip()
            amount_str = record[amount_col].strip()
            note = record[note_col].strip() if note_col >= 0 and len(record) > note_col else ""

            if not date_str or not payee or not amount_str:
                continue

            # Parse date
            date = None
            formats = [args.date_format, "%d/%m/%Y", "%Y/%m/%d", "%m/%d/%Y", "%d-%m-%Y", "%Y-%m-%d", "%m-%d-%Y"]
            for fmt in formats:
                try:
                    date = datetime.strptime(date_str, fmt)
                    break
                except:
                    continue

            if date is None:
                continue

            # Parse amount
            amount_str_clean = re.sub(r'[^\d\.,\-]', '', amount_str).replace(',', '.')
            try:
                amount = float(amount_str_clean)
            except ValueError:
                continue

            amount *= args.scale
            if args.neg:
                amount = -amount

            csv_amount = -amount
            expense_amount = -csv_amount

            # Classify payee
            classified_account = classify_payee(payee, classes, datas)

            # Print transaction - exactly as PHP
            if note:
                print(f";{note}")  # PHP doesn't add space after semicolon
            print(f"{date.strftime(TRANSACTION_DATE_FORMAT)} {payee}")

            # Print account lines with exact PHP formatting
            print(print_account_line(dest_account, csv_amount, args.columns))
            print(print_account_line(classified_account, expense_amount, args.columns))
            print()

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
